Source/Core/Core/PowerPC/GDBThread.cpp

```diff
@@ -0,0 +1,1274 @@
+    break;
+  case 'G':
+    GDBWriteRegisters();
+    break;
+  case 'p':
+    GDBReadRegister();
+    break;
+  case 'P':
+    GDBWriteRegister();
+    break;
+  case 'm':
+    GDBReadMemory();
+    break;
+  case 'M':
+    GDBWriteMemory();
+    break;
+  case 'c':
+    GDBContinue();
+    break;
+  case 's':
+    GDBStep();
+    break;
+  case ' ':
+    GDBPause();
+    break;
+  case 'z':
+    GDBRemoveBreakpoint();
+    break;
+  case 'Z':
+    GDBAddBreakpoint();
+    break;
+  default:
+    GDBAck();
+    GDBReply("");
+    break;
+  }
+}
+
+u8 GDBThread::GDBReadByte()
+{
+  if (!m_connected)
+    return 0;
+
+  u8 c;
+
+  size_t res = recv(m_sock, reinterpret_cast<char*>(&c), 1, MSG_WAITALL);
+  if (res != 1)
+    failr("recv failed");
+
+  return c;
+}
+
+u8 GDBThread::GDBCalculateChecksum() const
+{
+  u32 len = m_cmd_len;
+  const u8* ptr = m_cmd_bfr;
+  u8 c = 0;
+
+  while (len-- > 0)
+    c += *ptr++;
+
+  return c;
+}
+
+void GDBThread::GDBReply(const char* reply)
+{
+  if (!m_connected)
+    return;
+
+  memset(m_cmd_bfr, 0, sizeof m_cmd_bfr);
+
+  m_cmd_len = static_cast<u32>(strlen(reply));
+  if (m_cmd_len + 4 > sizeof m_cmd_bfr)
+    fail("GDBReply: cmd_bfr overflow");
+
+  memcpy(m_cmd_bfr + 1, reply, m_cmd_len);
+
+  m_cmd_len++;
+  u8 chk = GDBCalculateChecksum();
+  m_cmd_len--;
+  m_cmd_bfr[0] = GDB_STUB_START;
+  m_cmd_bfr[m_cmd_len + 1] = GDB_STUB_END;
+  m_cmd_bfr[m_cmd_len + 2] = NibbleToHex(chk >> 4);
+  m_cmd_bfr[m_cmd_len + 3] = NibbleToHex(chk);
+
+  const u8* ptr = m_cmd_bfr;
+  u32 left = m_cmd_len + 4;
+  while (left > 0)
+  {
+    int n = send(m_sock, reinterpret_cast<const char*>(ptr), left, 0);
+    if (n < 0)
+      fail("gdb: send failed");
+    left -= n;
+    ptr += n;
+  }
+}
+
+void GDBThread::GDBNak()
+{
+  if (!m_connected)
+    return;
+
+  const char nak = GDB_STUB_NAK;
+
+  size_t res = send(m_sock, &nak, 1, 0);
+  if (res != 1)
+    fail("send failed");
+}
+
+void GDBThread::GDBAck()
+{
+  if (!m_connected)
+    return;
+
+  const char ack = GDB_STUB_ACK;
+
+  size_t res = send(m_sock, &ack, 1, 0);
+  if (res != 1)
+    fail("send failed");
+}
+
+void GDBThread::GDBHandleSignal()
+{
+  if (!m_connected)
+    return;
+
+  char bfr[128] = {};
+
+  GDBAck();
+
+  switch (m_last_signal_condition)
+  {
+  case MemCheckCondition::Execute:
+    sprintf(bfr, "T%02X%08X:%08X", static_cast<int>(m_last_signal), 64, PC);
+    break;
+  case MemCheckCondition::Read:
+    sprintf(bfr, "T%02X%08X:%08X;rwatch:%08llX", static_cast<int>(m_last_signal), 64, PC,
+            m_last_signal_address);
+    break;
+  case MemCheckCondition::Write:
+    sprintf(bfr, "T%02X%08X:%08X;watch:%08llX", static_cast<int>(m_last_signal), 64, PC,
+            m_last_signal_address);
+    break;
+  case MemCheckCondition::ReadWrite:
+    sprintf(bfr, "T%02X%08X:%08X;awatch:%08llX", static_cast<int>(m_last_signal), 64, PC,
+            m_last_signal_address);
+    break;
+  default:
+    return;
+    break;
```

Comment:
You might want to remove this `break`, since some compiler might mark it as unreachable code after that `return`

Reply:
Makes sense.